cli/src/complete.rs

```diff
@@ -329,35 +332,88 @@ fn revisions(revisions: Option<&str>) -> Vec<CompletionCandidate> {
         let mut symbol_names: Vec<_> = revset_aliases.symbol_names().collect();
         symbol_names.sort();
-        candidates.extend(symbol_names.into_iter().map(|symbol| {
-            let (_, defn) = revset_aliases.get_symbol(symbol).unwrap();
-            CompletionCandidate::new(symbol)
-                .help(Some(defn.into()))
-                .display_order(Some(REVSET_ALIAS))
-        }));
+        candidates.extend(
+            symbol_names
+                .into_iter()
+                .filter(|symbol| symbol.starts_with(match_prefix))
+                .map(|symbol| {
+                    let (_, defn) = revset_aliases.get_symbol(symbol).unwrap();
+                    CompletionCandidate::new(symbol)
+                        .help(Some(defn.into()))
+                        .display_order(Some(REVSET_ALIAS))
+                }),
+        );
 
         Ok(candidates)
     })
 }
 
-pub fn mutable_revisions() -> Vec<CompletionCandidate> {
-    revisions(Some("mutable()"))
+fn revset_expression(
+    current: &std::ffi::OsStr,
+    revisions_filter: Option<&str>,
+) -> Vec<CompletionCandidate> {
+    let Some(current) = current.to_str() else {
+        return Vec::new();
+    };
+    let (prepend, match_prefix) = split_revset_trailing_name(current).unwrap_or(("", current));
+    let candidates = revisions(revisions_filter, match_prefix);
+    if prepend.is_empty() {
+        candidates
+    } else {
+        candidates
+            .into_iter()
+            .map(|candidate| candidate.add_prefix(prepend))
+            .collect()
+    }
 }
 
-pub fn all_revisions() -> Vec<CompletionCandidate> {
-    revisions(None)
+pub fn revset_expression_all(current: &std::ffi::OsStr) -> Vec<CompletionCandidate> {
+    revset_expression(current, None)
+}
+
+pub fn revset_expression_mutable(current: &std::ffi::OsStr) -> Vec<CompletionCandidate> {
+    revset_expression(current, Some("mutable()"))
+}
+
+/// Identifies if an incomplete expression ends with a name, or may be continued
+/// with a name.
+///
+/// If the expression ends with an name or a partial name, returns a tuple that
+/// splits the string at the point the name starts.
+/// If the expression is empty or ends with a prefix or infix operator that
+/// could plausibly be followed by a name, returns a tuple where the first
+/// item is the entire input string, and the second item is empty.
+/// Otherwise, returns `None`.
+///
+/// The input expression may be incomplete (e.g. missing closing parentheses),
+/// and the ability to reject invalid expressions is limited.
+pub fn split_revset_trailing_name(incomplete_revset_str: &str) -> Option<(&str, &str)> {
+    if incomplete_revset_str.trim_ascii().is_empty() {
+        return Some((incomplete_revset_str, ""));
+    }
+    let re = regex::Regex::new(
+        r"(?:^|[:~\|&\(,]|\.\.)\s*((?:[\p{XID_CONTINUE}_/]+[@.+-])*[\p{XID_CONTINUE}_/]*)$",
+    )
+    .unwrap();
```

Comment:
This pattern is hard to read. Can you switch to the verbose `(?x)` syntax?

It might also be better to first do `.rsplit_once()` by a known set of infix/prefix operators.

Reply:
I had to do two splits because `..` couldn't fit with the others, but the regex got dramatically simpler in the process, so I suppose the need for `(?x)` is reduced.